core/src/replay_stage.rs

```diff
@@ -644,72 +666,145 @@ impl ReplayStage {
+                stats.has_voted = tower.has_voted(bank.slot());
+                stats.is_recent = tower.is_recent(bank.slot());
+                if let Some(fp) = progress.get_mut(&bank.slot()) {
+                    fp.fork_stats = stats.clone();
+                }
+                stats
             })
             .collect();
+        let mut votable: Vec<_> = frozen_banks
+            .iter()
+            .zip(stats.iter())
+            .filter(|(_, stats)| stats.is_recent && !stats.has_voted && stats.vote_threshold)
+            .collect();
 
-        votable.sort_by_key(|b| b.0);
-        let ms = timing::duration_as_ms(&tower_start.elapsed());
+        //highest weight, lowest slot first
+        votable.sort_by_key(|b| (b.1.weight, 0i64 - b.1.slot as i64));
 
+        votable.iter().for_each(|(_, stats)| {
+            let mut parents: Vec<_> = if let Some(set) = ancestors.get(&stats.slot) {
+                set.iter().collect()
+            } else {
+                vec![]
+            };
+            parents.sort();
+            debug!("{}: {:?} {:?}", stats.slot, stats, parents,);
+        });
         trace!("votable_banks {}", votable.len());
-        if !votable.is_empty() {
-            let weights: Vec<u128> = votable.iter().map(|x| x.0).collect();
-            info!(
-                "@{:?} tower duration: {:?} len: {} weights: {:?}",
-                timing::timestamp(),
-                ms,
-                votable.len(),
-                weights
-            );
-        }
+        let rv = Self::pick_best_fork(ancestors, &votable);
+        let ms = timing::duration_as_ms(&tower_start.elapsed());
+        let weights: Vec<(u128, u64, u64)> = votable
+            .iter()
+            .map(|x| (x.1.weight, x.1.slot, x.1.block_height))
+            .collect();
+        debug!(
+            "@{:?} tower duration: {:?} len: {}/{} weights: {:?} voting: {}",
+            timing::timestamp(),
+            ms,
+            votable.len(),
+            stats.iter().filter(|s| !s.has_voted).count(),
+            weights,
+            rv.0.is_some()
+        );
         inc_new_counter_info!("replay_stage-tower_duration", ms as usize);
+        rv
+    }
 
-        votable
+    fn pick_best_fork(
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        best_banks: &[(&Arc<Bank>, &ForkStats)],
+    ) -> (Option<(Arc<Bank>, u64)>, Option<Arc<Bank>>) {
+        if best_banks.is_empty() {
+            return (None, None);
+        }
+        let mut rv = None;
+        let (best_bank, best_stats) = best_banks.last().unwrap();
+        debug!("best bank: {:?}", best_stats);
+        let mut by_slot: Vec<_> = best_banks.iter().collect();
+        by_slot.sort_by_key(|x| x.1.slot);
+        //look for the oldest ancestors of the best bank
+        if let Some(best_ancestors) = ancestors.get(&best_stats.slot) {
+            for (parent, parent_stats) in by_slot.iter() {
+                if parent_stats.is_locked_out {
```

Comment:
You want to set the "heaviest" (second item in the tuple) even if the bank is locked out so we can reset PoH recorder right?